src/Log/Log.php

```diff
@@ -350,7 +350,7 @@ public static function engine(string $name): ?LoggerInterface
      * @return bool Success
      * @throws \InvalidArgumentException If invalid level is passed.
      */
-    public static function write($level, $message, $context = []): bool
+    public static function write($level, string $message, $context = []): bool
```

Comment:
Adding typehints on stable releases technically is a BC break, as any class extending Log would break with a fatal error.
We could either make this a minor (4.next) or say that this is OK for now since we are early in 4.0.

Reply:
I think we have to commit this as it's a bug fix. Unfortunately, the bug is the missing type hint.

Reply:
That is not completely true, though. The fix is the docblock itself. The added typehint is an additional typehint that would break BC.
We can decide to do that, but it would not be necessary for the fix itself to be complete or working (and having 0 BC break).

Reply:
Up to you. With the existing code, we're giving the user broken api because of https://github.com/cakephp/cakephp/commit/fa745835a795ea97995b026bd7494802c24fcc37.

Reply:
You misread. The code itself is already fixed without the typehint (as it would already allow string to work fine). There is no need for the typehint here other than making it "cleaner". Functionally it would work both ways, only that the typehint breaks the public API in a BC breaking way.

Reply:
Concrete: You break the whole ecosystem that uses Log::write() as shown e.g. here: https://github.com/dereuromark/CakePHP-DatabaseLog/blob/master/src/Log/Engine/DatabaseLog.php#L49

Reply:
It is a big change come in post release.

Reply:
I think changing this typehint will also cause conflicts with the PSR interfaces.

Reply:
> Adding typehints on stable releases technically is a BC break, as any class extending Log::write() would break with a fatal error.

There is practically zero chance of someone using extended version of `Cake\Log\Log`. Lack of typehint here will simply cause an error at engine level.

> I think changing this typehint will also cause conflicts with the PSR interfaces.

No it doesn't. As you see the test suite is passing. `Cake\Log\Log.php` and `Cake\Log\LogTrait` are our extra classes. Only the engine under `Cake\Log\Engine` use the PSR interface and their `log()` methods still have exact same signature as that of `LogInterface::log()`.